### PR TITLE
Service Discovery ACLs

### PR DESCRIPTION
**Pull request by ryanuber**
Implements a filter on top of the state store to allow restricting read access to services. This allows issuing tokens which can only see services they have explicit permissions to read. The documentation is purposely left kind of half-complete so that we can polish it off after the DNS side of this is more complete, but we can tackle that in a separate PR.

### CODE REVIEW

File: consul/catalog_endpoint.go [hunk 1]
```diff
@@ -126,7 +126,7 @@ func (c *Catalog) ListNodes(args *structs.DCSpecificRequest, reply *structs.Inde
 		state.QueryTables("Nodes"),
 		func() error {
 			reply.Index, reply.Nodes = state.Nodes()
-			return nil
+			return c.srv.filterACL(args.Token, reply)
```
Comment (armon) [inline consul/catalog_endpoint.go]: Do nodes need to be filtered?

File: website/source/docs/internals/acl.html.markdown [hunk 2]
```diff
@@ -208,3 +211,22 @@ methods of configuring ACL tokens to use for registration events:
...
+filter out all services which the token has no access to in all API queries,
+making it appear as though the restricted services do not exist.
+
+Consul's DNS interface is also affected by restrictions to service
+registrations. If the token used by the agent does not have access to a given
+service, then the DNS interface will return no records when queried for it. If
+the [acl_default_policy](/docs/agent/options.html#acl_default_policy) is set to
+deny, this means that Consul will not be able to serve any DNS records that the
+[acl_token](/docs/agent/options.html#acl_token) is not explicitly granted read
+access to.
```
Comment (armon) [inline website/source/docs/internals/acl.html.markdown]: Should we add a node about the `anonymous` token to allow non-auth'd DNS?

**Comment by armon** [comment]
Minor comments, but LGTM! :muscle: :+1:

File: consul/catalog_endpoint.go (hunk 1 above)
Comment (ryanuber) [inline consul/catalog_endpoint.go]: Oh probably not, haha good catch.

File: website/source/docs/internals/acl.html.markdown (hunk 2 above)
Comment (ryanuber) [inline website/source/docs/internals/acl.html.markdown]: Yep makes sense. Probably pretty important actually.

File: consul/acl.go [hunk 3]
```diff
@@ -192,3 +194,168 @@ func (s *Server) useACLPolicy(id, authDC string, cached *aclCacheEntry, p *struc
...
+
+	case *structs.IndexedNodeServices:
+		filt.filterNodeServices(v.NodeServices)
+
+	case *structs.IndexedCheckServiceNodes:
+		filt.filterCheckServiceNodes(&v.Nodes)
+
+	case *structs.IndexedNodeDump:
+		filt.filterNodeDump(&v.Dump)
+	}
```
Comment (slackpad) [inline consul/acl.go]: This seems like the kind of thing that should freak out if a new type is encountered (we could add `case` clauses for ones that we explicitly don't care to filter).

Comment (armon) [inline consul/acl.go]: Ah yeah, like a default: panic

Comment (ryanuber) [inline consul/acl.go]: Seems reasonable, will add that in!

**Pull request merged**